### ISSUE TITLE
eth_estimateGas fails on empty address

### ISSUE THREAD

**Issue by kuegi** [DeFiCh/ain#2804]
## Summary

eth_estimateGas returns `Custom error: gas required exceeds allowance 0` if the address has no DFI on it. But it should return the estimated gas.

**Comment by kuegi** [comment]
thx. @sieniven it might be according to geth. but reality for exchanges is different it seems.
doing the call for any ethereum rpc on https://ethereumnodes.com/ with the address that is also empty on ETH, it always returns the estimated gas.
So this difference makes it harder for anyone to reuse existing eth tools.

**Comment by sieniven** [comment]
@kuegi that is not true. Please verify again on your end that you are directly interacting with an ethereum node. With any ethereum node, the default behaviour returns an invalid response: `"error": { "code": -32000, "message": "insufficient funds for transfer" }`. Often times these public endpoints, for example Alchemy, may provide an additional layer on top of nodes which differ from behaviours. We choose to align ourselves as much as we can to geth, and implement behaviours that makes sense in our context.

In addition, as mentioned above, please note that our node supports the addition state override set customization in our eth calls RPC. In this specific context, it does not make sense to return a default gas transfer used in this RPC when the address is zero. That makes the RPC faulty to begin with. Please pass the state override if you would like to set a defined context on your RPC call.

**Comment by LXHLeaner** [comment]
@sieniven I know  need dfi for gas fee.， We are using the eth_estimateGas interface to estimate the needed fee before transferring the DFI

**Comment by LXHLeaner** [comment]
@sieniven Account amount is not enough will also report an error?
![image](https://github.com/DeFiCh/ain/assets/93808747/cd81c6e5-7ddb-409a-9e44-177d3f5fe348)

**Comment by sieniven** [comment]
@LXHLeaner Yup. https://www.quicknode.com/docs/ethereum/eth_estimateGas - you can check out the optional state override parameter that our node supports.

**Comment by sieniven** [comment]
@kuegi 

> If I do the same call on any public eth rpc that I find, I get a valid result. So its a bit hard to believe that this should not be the standard.

 https://ethereum-sepolia.publicnode.com public RPC endpoint behaviour aligns exactly the same as our node.
 
> What can we do, to make it easier for exchanges to integrate DMC?

We have been making multiple PRs and fixes (more incoming) to fix all endpoints to align ourselves as much as we can with geth. geth is the most largest client with the most widespread user base. Actually the node has been making active steps to stability and full integration (PRs #2633, #2736, #2784, #2788, etc). Would be really helpful if the community could continue pointing out any integration issues with external web3 applications :)

**Comment by kuegi** [comment]
> @kuegi
> 
> > If I do the same call on any public eth rpc that I find, I get a valid result. So its a bit hard to believe that this should not be the standard.
> 
> https://ethereum-sepolia.publicnode.com public RPC endpoint behaviour aligns exactly the same as our node.

Does this call return something else on your side? If I call it, it returns a valid gas estimate:

```curl  https://ethereum-sepolia.publicnode.com \
-H 'Content-Type: application/json' \
--data '{
    "jsonrpc":"2.0","method":"eth_estimateGas",
    "params":[
        {
            "from":"0x4790b73359B393a10D4201371563795001f17977",
            "to":"0x347D48fB3907B71089945Aee7cafEe75d5F30EBc",
            "data":"0xa9059cbb000000000000000000000000dac0ebb0b5ed7ffd9441509acb40f06a529500000000000000000000000000000000000000000000000107ad8f556c6c0000"
        }],
        "id":1
}'```

returns `{"jsonrpc":"2.0","id":1,"result":"0x5512"}`

> Would be really helpful if the community could continue pointing out any integration issues with external web3 applications :)

Thats what I am trying to do here.

**Comment by sieniven** [comment]
@kuegi Thanks for providing the context. This is much more useful, and this is indeed a bug. Will be re-opening this issue.

**Comment by Stonygan** [comment]
> https://ethereum-sepolia.publicnode.com public RPC endpoint behaviour aligns exactly the same as our node.


I run a public BSC Node without some additional layers and Version: `instance=Geth/v1.3.8-c6aeee20-20240123/linux-amd64/go1.21.6`
Have the same behavior, Request works and result is the estimate Gas as expected, only Defichainnode returns an incorrect result.

**Comment by LXHLeaner** [comment]
@sieniven @Stonygan @ldenman @kuegi When will this problem be solved?

**Comment by sieniven** [comment]
This issue will be resolved once the PR #2808 is merged. The eth_estimateGas algorithm now fully aligns with geth's pipeline.
